File: JavaApp/src/main/java/com/charlezz/javaapp/di/AppModule.java
package com.charlezz.javaapp.di;

import android.app.Application;
import android.content.ContentValues;
import android.content.Context;
import android.database.sqlite.SQLiteDatabase;

import androidx.annotation.NonNull;
import androidx.room.Room;
import androidx.room.RoomDatabase;
import androidx.sqlite.db.SupportSQLiteDatabase;

import com.charlezz.javaapp.App;
import com.charlezz.javaapp.feature.local.AppDatabase;
import com.charlezz.javaapp.feature.local.User;
import com.charlezz.javaapp.feature.local.UserDao;
import com.charlezz.javaapp.feature.remote.PostService;

import java.util.Locale;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

@Module(includes = ViewModelModule.class)
public class AppModule {

    @Provides
    @Singleton
    Application bindApp(App app){
        return app;
    }

    @Provides
    @Singleton
    PostService providePostService(){
        return new Retrofit.Builder()
                .baseUrl("https://jsonplaceholder.typicode.com/")
                .addConverterFactory(GsonConverterFactory.create())
                .build()
                .create(PostService.class);
    }

    @Provides
    @Singleton
    Context provideContext(App app){
        return app;
    }

    @Provides
    @Singleton
    AppDatabase provideAppDatabase(Context context){
        return Room
                .inMemoryDatabaseBuilder(context,AppDatabase.class) // temporary
//                .databaseBuilder(context, AppDatabase::class.java, AppConstants.DB_NAME) // permanent
                .fallbackToDestructiveMigration()
                .addCallback(new RoomDatabase.Callback() {
                    @Override
                    public void onCreate(@NonNull SupportSQLiteDatabase db) {
                        super.onCreate(db);
                        for( int i =0; i< 10000;i++){
                            ContentValues cv = new ContentValues();
                            cv.put("name", String.format(Locale.getDefault(),"Name %d", i));
                            cv.put("birth" ,String.format(Locale.getDefault(),"Birth %d", i));
                            db.insert(User.class.getSimpleName(), SQLiteDatabase.CONFLICT_REPLACE,cv);
                        }
                    }
                })
                .build();
    }

    @Provides
    @Singleton
    UserDao providePersonDao(AppDatabase appDatabase){
        return appDatabase.dao();
    }

}

File: KotlinApp/src/main/java/com/charlezz/finalarchitecture/di/AppModule.kt
package com.charlezz.finalarchitecture.di

import android.app.Application
import androidx.sqlite.db.SupportSQLiteDatabase
import androidx.room.Room
import androidx.room.RoomDatabase
import android.content.ContentValues
import android.content.Context
import android.database.sqlite.SQLiteDatabase
import com.charlezz.finalarchitecture.App
import com.charlezz.finalarchitecture.AppConstants
import com.charlezz.finalarchitecture.feature.local.AppDatabase
import com.charlezz.finalarchitecture.feature.local.DBHelper
import com.charlezz.finalarchitecture.feature.local.DBHelperImpl
import com.charlezz.finalarchitecture.feature.local.PersonDao
import com.charlezz.finalarchitecture.feature.photo.PhotoHelper
import com.charlezz.finalarchitecture.feature.photo.PhotoHelperImpl
import com.charlezz.finalarchitecture.feature.pref.PreferencesHelper
import com.charlezz.finalarchitecture.feature.pref.PreferencesHelperImpl
import com.charlezz.finalarchitecture.feature.remote.ApiHelper
import dagger.Module
import dagger.Provides
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import javax.inject.Singleton

@Module
class AppModule {

    @Provides
    @Singleton
    fun provideAppContext(app: App): Context {
        return app
    }

    @Provides
    @Singleton
    fun provideApplication(app: App): Application {
        return app
    }

    @Provides
    @Singleton
    fun provideAppDatabase(context: Context): AppDatabase {
        return Room
                .inMemoryDatabaseBuilder(context, AppDatabase::class.java) // temporary
//                .databaseBuilder(context, AppDatabase::class.java, AppConstants.DB_NAME) // permanent
                .fallbackToDestructiveMigration()
                .addCallback(object : RoomDatabase.Callback(){
                    override fun onCreate(db: SupportSQLiteDatabase) {
                        super.onCreate(db)
                        for( index in 0 until 10000){
                            val cv = ContentValues()
                            cv.put("name", "Name$index")
                            cv.put("birth", "$index")
                            db.insert("person", SQLiteDatabase.CONFLICT_REPLACE,cv)
                        }
                    }
                })
                .build()
    }

    @Provides
    @Singleton
    fun provideApiHelper(): ApiHelper = Retrofit.Builder()
            .baseUrl("https://jsonplaceholder.typicode.com/")
            .addConverterFactory(GsonConverterFactory.create())
            .build()
            .create(ApiHelper::class.java)

    @Provides
    @Singleton
    fun providePersonDao(appDatabase: AppDatabase): PersonDao {
        return appDatabase.dao()
    }

    @Provides
    @Singleton
    fun provideDataManager(personDao: PersonDao): DBHelper {
        return DBHelperImpl(personDao)
    }

    @Provides
    @Singleton
    fun providePreferencesHelper(context:Context): PreferencesHelper {
        return PreferencesHelperImpl(context, AppConstants.PREF_NAME)
    }

    @Provides
    @Singleton
    fun providePhotoHelper(): PhotoHelper {
        return PhotoHelperImpl()
    }
}